model-client/src/jvmMain/kotlin/org/modelix/model/oauth/ModelixOAuthClient.kt

```diff
@@ -66,42 +66,27 @@ object ModelixOAuthClient {
         }
     }
 
-    fun installAuth(config: HttpClientConfig<*>, baseUrl: String, authTokenProvider: (() -> String?)? = null) {
+    actual fun installAuth(config: HttpClientConfig<*>, baseUrl: String, authTokenProvider: (suspend () -> String?)?) {
+        if (authTokenProvider != null) {
+            installAuthWithAuthTokenProvider(config, authTokenProvider)
+        } else {
+            installAuthWithPKCEFlow(config, baseUrl)
+        }
+    }
+
+    private fun installAuthWithPKCEFlow(config: HttpClientConfig<*>, baseUrl: String) {
         config.apply {
             install(Auth) {
                 bearer {
                     loadTokens {
-                        val tp = authTokenProvider
-                        if (tp == null) {
-                            ModelixOAuthClient.getTokens()?.let { BearerTokens(it.accessToken, it.refreshToken) }
-                        } else {
-                            val token = tp()
-                            if (token == null) {
-//                            connectionStatus = RestWebModelClient.ConnectionStatus.WAITING_FOR_TOKEN
-                                null
-                            } else {
-                                BearerTokens(token, "")
-                            }
-                        }
+                        getTokens()?.let { BearerTokens(it.accessToken, it.refreshToken) }
                     }
                     refreshTokens {
-                        val tp = authTokenProvider
-                        if (tp == null) {
-                            var url = baseUrl
-                            if (!url.endsWith("/")) url += "/"
-                            if (url.endsWith("/model/")) url = url.substringBeforeLast("/model/")
-//                        connectionStatus = RestWebModelClient.ConnectionStatus.WAITING_FOR_TOKEN
-                            val tokens = ModelixOAuthClient.authorize(url)
-                            BearerTokens(tokens.accessToken, tokens.refreshToken)
-                        } else {
-                            val providedToken = tp()
-                            if (providedToken != null && providedToken != this.oldTokens?.accessToken) {
-                                BearerTokens(providedToken, "")
-                            } else {
-//                            connectionStatus = RestWebModelClient.ConnectionStatus.WAITING_FOR_TOKEN
-                                null
-                            }
-                        }
+                        var url = baseUrl
+                        if (!url.endsWith("/")) url += "/"
+                        if (url.endsWith("/model/")) url = url.substringBeforeLast("/model/")
```

Comment:
What is this magic? Can you document this now that this is touched?

Reply:
Done.
Commented it and created a ticket.

Reply:
Yikes, that is really a hack.